### ISSUE TITLE
Speed on new fairseq is 1/12 of the transformer example

### ISSUE THREAD

**Issue by kkissmart** [pytorch/xla#2346]
## ❓ Questions and Help
I am running the latest fairseq code (https://github.com/pytorch/fairseq) using transformer seq2seq model ( I changed train.py and trainer to make it work on tpu), but the word per second is only 500 on each core. When I follow this example https://cloud.google.com/tpu/docs/tutorials/transformer-pytorch, the word per second is 6000. Do I miss anything here?

**Comment by taylanbil** [comment]
Hello @kkissmart,

A bit of context: TPU support was recently added to the upstream `fairseq` via https://github.com/pytorch/fairseq/pull/2245. In the tutorial, we use our fork of `fairseq`, which is based on an older upstream commit at the moment. The goal is to merge the two and make sure the upstream branch works just as well as the fork, and we intend to look at this in the near future, hopefully.

Back to your issue; without seeing your code changes, a command line, or [a metrics_report](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#get-a-metrics-report), this is tough to diagnose, but my wild guess is that you are probably suffering from dynamic shape inputs. If this is the case, it would manifest itself in the metrics report as a high number (and time spent) of compilations. Are you using the [`--num-batch-buckets` flag](https://github.com/pytorch/fairseq/pull/2245/commits/851d456152a8f919c23818b456b458cf1c982701) and setting it to a non-zero number?

If this is not the case, could you please provide a bit more info? A command line, a diff showing your code changes, and a metrics report?

**Comment by kkissmart** [comment]
@taylanbil Thanks so much!

Is there a readme for https://github.com/pytorch/fairseq/pull/2245 on how to run fairseq on TPU efficiently?

Thanks a lot!

**Comment by taylanbil** [comment]
Ah that I don't know, paging @myleott on the question of whether there are docs on fairseq upstream and tpus.

Is your issue resolved @kkissmart ?

**Comment by kkissmart** [comment]
@taylanbil no. it is still very very slow. I pulled the  fairseq code today and made no changes.

here is the command I have used

python train.py $HOME/pytorch-tutorial-data/wmt18_en_de_bpej32k   --save-interval=1   --arch=transformer_vaswani_wmt_en_de_big --attention-dropout=0.1   --no-progress-bar   --criterion=label_smoothed_cross_entropy   --source-lang=en   --lr-scheduler=inverse_sqrt   --min-lr 1e-09   --skip-invalid-size-inputs-valid-test   --target-lang=de   --label-smoothing=0.1   --update-freq=1   --optimizer adam   --adam-betas '(0.9, 0.98)'   --warmup-init-lr 1e-07   --lr 0.0005   --warmup-updates 4000   --share-all-embeddings   --dropout 0.3   --weight-decay 0.0   --valid-subset=valid   --max-epoch=25 --tpu --bf16 --max-tokens 4096 --num-batch-buckets 8 --distributed-world-size 8

My logs:
2020-07-14 05:41:08 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:45:25 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:46:24 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:46:27 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:46:28 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:48:05 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:48:06 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:49:45 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:51:22 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:53:10 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:54:11 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:54:55 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:55:01 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:55:50 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:55:52 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:55:59 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:56:02 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:56:07 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:56:19 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:56:38 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:56:38 | INFO | train_inner | epoch 001:    100 / 9894 loss=13.127, nll_loss=12.807, ppl=7168.1, wps=0, ups=0, wpb=14678, bsz=352, num_updates=100, lr=1.25975e-05, gnorm=1.263, train_wall=451, wall=1053
2020-07-14 05:56:38 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:56:44 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:57:34 | INFO | train_inner | epoch 001:    200 / 9894 loss=12.179, nll_loss=11.734, ppl=3405.85, wps=249.6, ups=0.02, wpb=14020, bsz=552, num_updates=200, lr=2.5095e-05, gnorm=1.893, train_wall=21, wall=1109
2020-07-14 05:57:34 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 05:58:27 | INFO | train_inner | epoch 001:    300 / 9894 loss=11.616, nll_loss=11.044, ppl=2111.69, wps=296.6, ups=0.02, wpb=15764, bsz=464, num_updates=300, lr=3.75925e-05, gnorm=3.354, train_wall=20, wall=1162
2020-07-14 05:58:28 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning

The Word per second here is too low. 

Thanks for the help.

**Comment by myleott** [comment]
Sorry, the reported words per second on TPUs are misreported. If your log-interval is 100, then the wps is actually 100 times larger than what's reported. So the actual words per second on the last line is 40160.

> Ah that I don't know, paging @myleott on the question of whether there are docs on fairseq upstream and tpus.

No, we haven't added docs yet but will soon, starting with RoBERTa (since that's been full reproduced on TPUs) and then translation.

**Comment by taylanbil** [comment]
Thanks @myleott ! I forgot about that, so indeed this run seems to be on the same order of magnitude, speed-wise, to the tutorial (40k / 8 = 5k compared to 6k). @kkissmart you could experiment with `--max-tokens` to see if larger batches can give you a faster throughput. Also, if you add

```
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met

if i == 100:
    xm.master_print(met.metrics_report())
```

right after [this line](https://github.com/pytorch/fairseq/blob/master/fairseq_cli/train.py#L224), we could see if there are other factors that are slowing the process down following [our troubleshooting page](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#understand-the-metrics-report).

**Comment by kkissmart** [comment]
@taylanbil one more question before I close this issue. Usually I can fit in 8192 tokens on roberta base on single V100 32GB GPU, based on your experience, what is the max-token i can set for V3-8 here.

I have another issue actually. after I kill one training command, I usually have to restart the cpu instance to start another one. Do you have better solution for this?

**Comment by taylanbil** [comment]
TPU v3-8 devices have 16gb HBM. But it should be able to handle 128x64 inputs as in the tutorial, which is 8192 tokens. Enabling bf16 with the environment flag `XLA_USE_BF16=1`, we were able to go to 16k tokens as well. I'd say try w/ 8192 first.

Does "restart the cpu instance" mean restart the VM? If so, you shouldn't have to do that, what happens if you don't restart the VM? Maybe there are zombie python processes, you could kill with `pkill -9 python` or equivalent and retry?

**Comment by kkissmart** [comment]
@taylanbil  
16K tokens sound so amazing! 
If I don't restart VM, the next job hangs -- no error msg or progress
I tried pkill -9 python but it didnt work well.  
When I used GPU on Google Cloud, I need to do 1) pkill -9 python 2) sudo nvidia-smi --gpu-reset 
I was wondering if there is any other similar command for TPU, that I dont need to restart my VM every time.

Thanks very much!

**Comment by taylanbil** [comment]
> If I don't restart VM, the next job hangs -- no error msg or progress

That sounds weird, definitely should not be happening. I routinely fire a job after another one is done or interrupted (without restarting), and have no issues. Nothing gets printed to stdout/stderr at all? Command just sits there? How long do you wait before deciding it's hanging?

**Comment by taylanbil** [comment]
Closing this for now per the parallel thread under github/pytorch/fairseq, please re-open if issue happens.
